### PR TITLE
Feat: Usage Stats

### PR DESCRIPTION
**Pull request by Meldiron**
Usage stats are useful in Appwrite in future functions proxy. These stats will allow us to decide which function container to use or if we need to spawn a new one.

This implements `getStats()` that gives CPU, memory, disk, and network usage.

This is a per-container implementation of a similar implementation from utopia/system. Notice that the System library gives host machine stats while this gives per-container stats.

- [x] Implement tests

### CODE REVIEW

File: src/Orchestration/Adapter/DockerAPI.php [hunk 1]
```diff
@@ -278,6 +278,75 @@ public function networkDisconnect(string $container, string $network, bool $forc
...
+                $filtersSorted[$key] = [$value];
+            }
+    
+            $body = [
+                'all' => true,
+                'filters' => empty($filtersSorted) ? new stdClass() : json_encode($filtersSorted)
+            ];
+
+            $containers = \json_decode($this->call('http://localhost/containers/json?'.\http_build_query($body), 'GET')['response'], true);
+            $containerIds = \array_map(fn($c) => $c['Id'], $containers);
```
Comment (christyjacob4) [inline src/Orchestration/Adapter/DockerAPI.php:307]: Why aren't we using the `list` method of this adapter here ?
https://github.com/utopia-php/orchestration/blob/114046b9010ccb331deb7e23ef4aff22f903b890/src/Orchestration/Adapter/DockerAPI.php#L409

File: src/Orchestration/Adapter/DockerAPI.php [hunk 2]
```diff
@@ -278,6 +278,75 @@ public function networkDisconnect(string $container, string $network, bool $forc
...
+
+            $networkIn = 0;
+            $networkOut = 0;
+            foreach ($stats['networks'] as $network) {
+                $networkIn += $network['rx_bytes'];
+                $networkOut += $network['tx_bytes'];
+            }
+
+            // Rx means Receive, and Tx means
+
```
Comment (christyjacob4) [inline src/Orchestration/Adapter/DockerAPI.php:334]:
```suggestion
```

File: src/Orchestration/Adapter/DockerAPI.php [hunk 3]
```diff
@@ -278,6 +278,75 @@ public function networkDisconnect(string $container, string $network, bool $forc
...
+            $list[] = [
+                'id' => $stats['id'],
+                'name' => \ltrim($stats['name'], '/'), // Remove '/' prefix
+                'cpu' => 1, // TODO: Implement (I coudl not do it because Docker API does not give correct values)
+                // 'cpu' => ($cpuDelta / $systemCpuDelta) * $stats['cpu_stats']['online_cpus'] * 100.0,
```
Comment (christyjacob4) [inline src/Orchestration/Adapter/DockerAPI.php:339]: Can we clean this up a bit ?

File: src/Orchestration/Adapter/DockerAPI.php [hunk 4]
```diff
@@ -278,6 +278,75 @@ public function networkDisconnect(string $container, string $network, bool $forc
...
+                'memory' => ($stats['memory_stats']['usage'] / $stats['memory_stats']['limit']) * 100.0,
+                'diskIO' => [ 'in' => 0, 'out' => 0 ], // TODO: Implement (I could not find it in API)
+                'memoryIO' => [ 'in' => 0, 'out' => 0 ], // TODO: Implement (I could not find it in API)
```
Comment (christyjacob4) [inline src/Orchestration/Adapter/DockerAPI.php:342]:
```suggestion
                'diskIO' => [ 'in' => 0, 'out' => 0 ], // TODO: Implement (API does not provide these values)
                'memoryIO' => [ 'in' => 0, 'out' => 0 ], // TODO: Implement (API does not provide these values
)
```

File: src/Orchestration/Adapter/DockerCLI.php [hunk 5]
```diff
@@ -102,6 +102,135 @@ public function networkDisconnect(string $container, string $network, bool $forc
...
+                foreach($stdoutArray as $row) {
+                    if(empty($row)) {
+                        continue;
+                    }
+
+                    $rowParsed = [];
+                    \parse_str($row, $rowParsed);
+                    $containerIds[] = $rowParsed['id'];
+                }
+            }
```
Comment (christyjacob4) [inline src/Orchestration/Adapter/DockerCLI.php:146]: https://github.com/utopia-php/orchestration/blob/114046b9010ccb331deb7e23ef4aff22f903b890/src/Orchestration/Adapter/DockerCLI.php#L291

We can use the list function

File: src/Orchestration/Adapter/DockerCLI.php [hunk 6]
```diff
@@ -102,6 +102,135 @@ public function networkDisconnect(string $container, string $network, bool $forc
...
+                'diskIO' => $this->parseIOStats($stat['diskIO']),
+                'memoryIO' => $this->parseIOStats($stat['memoryIO']),
+                'networkIO' => $this->parseIOStats($stat['networkIO']),
+            ];
+        }
+
+        return $stats;
+    }
+
+    private function parseIOStats(string $stats) {
```
Comment (christyjacob4) [inline src/Orchestration/Adapter/DockerCLI.php:193]: Can we have a doc block explaining this function?

**Review by christyjacob4 — CHANGES_REQUESTED** [review]
Looks good. please address the comments

File: src/Orchestration/Adapter/DockerAPI.php (hunk 1 above)
Comment (Meldiron) [reply src/Orchestration/Adapter/DockerAPI.php:307]: Used `list` method instead.

File: src/Orchestration/Adapter/DockerAPI.php (hunk 3 above)
Comment (Meldiron) [reply src/Orchestration/Adapter/DockerAPI.php:337]: Removed comment. Formular should have worked, but didn't. So instead we stick with TODO.

File: src/Orchestration/Adapter/DockerCLI.php (hunk 5 above)
Comment (Meldiron) [reply src/Orchestration/Adapter/DockerCLI.php:146]: Used `list` instead.

File: src/Orchestration/Adapter/DockerCLI.php (hunk 6 above)
Comment (Meldiron) [reply src/Orchestration/Adapter/DockerCLI.php:193]: Added doc block.

File: src/Orchestration/Adapter/DockerCLI.php [hunk 7]
```diff
@@ -102,6 +102,119 @@ public function networkDisconnect(string $container, string $network, bool $forc
...
+        }
+
+        return $stats;
+    }
+
+
+     /**
+     * Use this method to parse string format into numeric in&out stats.
+     * CLI provides IO stats in verbose format like '12.5GiB / 10.01KiB'.
+     *
```
Comment (christyjacob4) [inline src/Orchestration/Adapter/DockerCLI.php:174]: Can you add an example input and output here?

File: src/Orchestration/Orchestration.php [hunk 8]
```diff
@@ -128,6 +129,91 @@ public function networkConnect(string $container, string $network): bool
         return $this->adapter->networkConnect($container, $network);
     }
 
+    /**
+     * Get usage stats of containers
```
Comment (eldadfux) [inline src/Orchestration/Orchestration.php:133]: Container or containers?

File: src/Orchestration/Orchestration.php [hunk 9]
```diff
@@ -128,6 +129,91 @@ public function networkConnect(string $container, string $network): bool
...
+     * @return array<Stats>
+     */
+    public function getStats(string $container = null, array $filters = [], int $cycles = 1): array 
+    {
+        /**
+         * @var array<Stats>
+         */
+        $averageStats = [];
+
+        for ($i = 0; $i < $cycles; $i++) {
```
Comment (eldadfux) [inline src/Orchestration/Orchestration.php:148]: Not sure why we calculate the average.

File: src/Orchestration/Orchestration.php (hunk 8 above)
Comment (Meldiron) [reply src/Orchestration/Orchestration.php:133]: Containers I think.

If you provide container ID, you get array with 1 Stat instance.

If you don't provide ID, you get array with Stat instance for each container that exists.

File: src/Orchestration/Orchestration.php (hunk 9 above)
Comment (eldadfux) [reply src/Orchestration/Orchestration.php:148]: Should we change the default from 1?

Comment (Meldiron) [reply src/Orchestration/Orchestration.php:148]: Changed default to 2.

**Review by christyjacob4 — APPROVED** [review]

**Review by eldadfux — APPROVED** [review]

**Review by christyjacob4 — APPROVED** [review]

**Pull request merged**